### ISSUE TITLE
Bulgarian Lev has wrong Currency Format 

### ISSUE THREAD

**Issue by keilw** [JavaMoney/jsr354-ri#151]
Migrated from https://java.net/jira/browse/JAVAMONEY-164

**Comment by keilw** [comment]
IMO a showstopper, too. Have to dig it up from the JIRA archive to see the history.

**Comment by atsticks** [comment]
By default the formats of the underlying Java platform are used. So the question is, if the format is invalid in the JDK's DecimalFormat as well...?

**Comment by keilw** [comment]
Wouldn't be the first time, see lack of Indian numberFormat support.

**Comment by atsticks** [comment]
For fixing it I have to know, what the exact/correct formatting should be? 

- Numeric format
- Symbol
- Currency format (numeric format + code/symbol).

Can anybody help here?

**Comment by Koziolek** [comment]
Valid format:
- digits in major coma separated, grouped by 3
- then coma as separator
- then currency symbol

Example:
1 123 000,50 лв.

**Comment by atsticks** [comment]
Number formatting depends on the country locale, so I propose to adapt the JDK pattern for the Locale(country="BG"). The corresponding test (working on my machine with unpushed changes) looks as follows:

```
MonetaryAmount money = Money.of(1123000.50, "BGN");
        Locale locale = new Locale("", "BG");
        MonetaryAmountFormat format = MonetaryFormats.getAmountFormat(
                AmountFormatQueryBuilder.of(locale).set(CurrencyStyle.SYMBOL)
                        .build());
        assertEquals(format.format(money), "1 123 000,50 лв");

        format = MonetaryFormats.getAmountFormat(
                AmountFormatQueryBuilder.of(locale).set(CurrencyStyle.CODE)
                        .build());
        assertEquals(format.format(money), "1 123 000,50 BGN");
```
Looks that OK for you?

**Comment by keilw** [comment]
Does it work for both RIs?

**Comment by stokito** [comment]
Sorry guys, but this was implemented incorrectly. The change to fix the issue is 1fa95fe271a5d2664e24b7711ff057c168cdbda0

First of all correct symbol for Bulgarian Levi is `лв.` not `лв` (i.e. missed a dot).
Secondly JDK actually works correctly but to see the `лв.` symbol you need to specify Bulgarian locale with both language and country i.e. `bg_BG`. Here is a snippet which shows the difference:
```java
public class CurrencyCode {
    public static void main(String[] args) {
        Locale country = new Locale("", "BG"); // lang will be ignored
        Currency currency = Currency.getInstance(country);
        Locale bulgarianLocale = new Locale("bg", "BG");
        String symbolForBg = currency.getSymbol(bulgarianLocale);
        String symbolForUs = currency.getSymbol(Locale.US);
        System.out.println("For country " + country.getCountry() + ", currency symbol is " + symbolForBg);
        System.out.println("For country " + Locale.US.getCountry() + ", currency symbol is " + symbolForUs);
    }
}
```
The result of execution:
```
For country BG, currency symbol is лв.
For country US, currency symbol is BGN
``

This actually makes sense because for other locales this might be confusing what the `лв.` means and they even can't read it if they don't know Cyrillic letters.
Also some symbols like `$` may be used in different countries to represent different currencies. In Honk Kong peoples on streets use `$` to represent Honk Kong Dollars. Here is a simple of test case:
```java
    @Test
    public void testPrint_SYMBOL_HKD() throws IOException {
        CurrencyToken token = new CurrencyToken(SYMBOL, new Locale("en", "HK"));
        FastMoney amount = FastMoney.of(0, "HKD");
        StringBuilder sb = new StringBuilder();
        token.print(sb, amount);
        assertEquals(sb.toString(), "HK$");
    }
```

So for the locale `en_HK` (and `zh_HK`) the symbol will be `HK$` and according to 
[wikipedia](https://en.wikipedia.org/wiki/Hong_Kong_dollar) this is one of the symbols which is used in banking.
So it may be a situation when few symbols for the one currency are in use.

But what is interesting that if you try to get symbol for USD with France locale `fr_FR` you'll get `$US` i.e. the dollar sign will be first and not last as in `HK$`. The same situation for GBP for french locale: it returns `£GB`. So we see here a pattern `SymbolCountry`.
But this not happens for some other currencies:
```
RUB for ru_RU: `₽`
RUB for en_US: `RUB`
UAH for uk_UA: `₴`
UAH for en_US: `UAH`
```

So  the resolving of currency symbol may be tricky.

Another problem that was fixed in 1fa95fe271a5d2664e24b7711ff057c168cdbda0 was parsing and formatting of thousands. 

Into `AmountNumberToken.initDecimalFormats()` method was added the following lines:
```java
        // Fix for https://github.com/JavaMoney/jsr354-ri/issues/151
        if ("BG".equals(locale.getCountry())) {
            formatFormat.setGroupingSize(3);
            formatFormat.setGroupingUsed(true);
            syms = formatFormat.getDecimalFormatSymbols();
            syms.setDecimalSeparator(',');
            syms.setGroupingSeparator(' ');
            formatFormat.setDecimalFormatSymbols(syms);
            parseFormat.setDecimalFormatSymbols(syms);
        }
```

In fact this all can be simplified to one line:
```java
            syms.setGroupingSeparator(' ');
```
I.e. by default the `groupingSeparator` is non-breakable space (NBSP) and this breaks code so here it's overridden with usual space.
Actually the same thing happens for other locales who using space for thousands separator.
So I would like to suggest more generic solution aab2d6586eeb930e64714a6d94338eeec59a4957 :

```java
       DecimalFormatSymbols syms = amountFormatContext.get(DecimalFormatSymbols.class);
        if (Objects.nonNull(syms)) {
            formatFormat.setDecimalFormatSymbols(syms);
            parseFormat.setDecimalFormatSymbols(syms);
        } else {
            syms = formatFormat.getDecimalFormatSymbols();
            if (syms.getGroupingSeparator() == NBSP) {
                syms.setGroupingSeparator(' ');
            }
            formatFormat.setDecimalFormatSymbols(syms);

            syms = parseFormat.getDecimalFormatSymbols();
            if (syms.getGroupingSeparator() == NBSP) {
                syms.setGroupingSeparator(' ');
            }
            parseFormat.setDecimalFormatSymbols(syms);
        }
```
I.e. if someone explicitly set `DecimalFormatSymbols` into context then use it. Otherwise get default DecimalFormatSymbols for the format and if it's  `groupingSeparator` is NBSP then override it with usual space.

**Comment by stokito** [comment]
Hi @Koziolek please review the 34cef7d commit. It breaks the change made previously
